---
title: "Data preprocessing test"
output: html_notebook
---

I want to see if the data can be preprocessed in exactly the same way but using the .rda files. It seems like a simpler way of doing things.

Setting up and file paths:
```{r}
library(tidyverse)
library(dplyr)
library(TCGAbiolinks)
library(SummarizedExperiment)

# you might have to setwd() (you want it set to data_preprocessing/)
print(getwd())

blacklist.file <- 'blacklist/Naeem2014_blacklist_annotation.txt'
```


Loading in the data:
```{r}

cancer.type.code <- 'BRCA'
load(paste(cancer.type.code, 'methylation.rda', sep='')) # will create a variable called 'data'
beta_values <- assay(data)
```


# Cleaning the data: (from cleaning_data.R)

Removing blacklisted files:
```{r}
# read blacklist into dataframe
blacklist <- read.csv(file = blacklist.file, header=TRUE, sep='\t') # has a column called 'Flag.discard.keep.' which either is 'discard' or 'keep'

blacklisted_probes <- blacklist %>%
                      filter(Flag.discard.keep. == 'discard') %>%
                      dplyr::select(probe)

beta_values_no_blacklisted <- beta_values[!rownames(beta_values) %in% blacklisted_probes$probe, ] # get all rows that are not in blacklisted probes

dim(beta_values_no_blacklisted)[1]
dim(beta_values)[1] - dim(blacklisted_probes)[1] # yay they are the same

```

Remove * chromosomes
No longer ordering it by chromosome, as that seems pointless.
```{r}
library(GenomicRanges)
probe_granges <- rowRanges(data)

# # sorted_probe_granges <- sort(probe_granges) %>% # the sort function orders by chromosome and start position, 485577 probes
#                         as_tibble() %>%
#                         filter(seqnames != "*") %>% # we also take out unknown (?) chromosomes, get 480457 probes
#                         filter(Composite.Element.REF %in% rownames(beta_values_no_blacklisted)) # getting the ones that were not blacklisted, 293808 probes



# not sorting:
sorted_probe_granges <- probe_granges %>%
                        as_tibble() %>%
                        filter(seqnames != "*") %>% # we also take out unknown (?) chromosomes, get 480457 probes
                        filter(Composite.Element.REF %in% rownames(beta_values_no_blacklisted)) # getting the ones that were not blacklisted, 293808 probes

beta_values_sorted <- beta_values_no_blacklisted[c(sorted_probe_granges$Composite.Element.REF), ] # selecting rows in order of sorted_probe_granges



# checking sizes:

dim(sorted_probe_granges) # don't have unknown chromosome probes and blacklisted probes 
dim(beta_values_no_blacklisted)
dim(beta_values_sorted)

d<- sort(probe_granges) %>% # number of probes on an unknown chromosome
                        as_tibble() %>%
                        filter(seqnames == "*") 

sum(rownames(beta_values_no_blacklisted) %in% d$Composite.Element.REF) # show some of these probes on unknown chomosome are removed from blacklisting


```

Remove probes with an na value fraction of more than 0.05
```{r}
na_frac <- is.na(beta_values_sorted) %>%
  as_tibble() %>%
  rowMeans()


beta_values_not_na <- beta_values_sorted[na_frac <= 0.05, ] # only keep probes with less than 0.05 na values

# getting the names of the probes removed:
na_probes_removed <- rownames(beta_values_sorted[na_frac > 0.05, ])

# Some of these row names look weird. Verifying that these also apear in the original beta values:
t <- function(probe) {
  if (substring(probe, 2, 2) == 'h') return(probe)
  else return("")
}
beta_values[unlist(map(rownames(beta_values), t)) != "", 1]

mean(is.na(beta_values_not_na)) # there are still some na values left

```



Impute (=estimate) the remaining na values:
```{r}
library(impute)

total_frac_na <- mean(is.na(beta_values_not_na))

imputed_betas <- impute.knn(beta_values_not_na, k=10, rowmax=0.25) # we allow any row to have max 25% missing data, no more
imputed_betas <- imputed_betas[[1]]


total_frac_na_imputed <- mean(is.na(imputed_betas)) # should be 0


```

Getting the m values:
```{r}
library(lumi)
m_values <- beta2m(imputed_betas)

# beta2m(0.000000000000000000000000000000000000000000000000000000000000000000000000000001)

```

Writing to file (both beta and m values):
```{r}

print(paste('PROJECT: ', cancer.type.code, ', WRITING M VALUES MATRIX ', sep = ''))

# set column and row names as it is for the result of the txt file processing (so it can be run by clean_data_preprocessing)
# imputed_betas <- cbind(probe=rownames(imputed_betas), imputed_betas) # adding probe name as a column
# rownames(imputed_betas) <- 1:nrow(imputed_betas)
# m_values <- cbind(probe=rownames(m_values), m_values) # adding probe name as a column
# rownames(m_values) <- 1:nrow(m_values)

write.csv(m_values, file=paste(cancer.type.code, '_m_from_rda', sep=''), sep='\t', row.names = TRUE, quote = FALSE)

# write.csv(imputed_betas, file=paste(save.path, cancer.type.code, '_beta_from_rda', sep=''), sep='\t', row.names = TRUE, quote = FALSE)


```

See EvaluateRdaProcessing for an evaluation of these results.













